### PR TITLE
feat: fix default retryer bug where initial wait period is the provid…

### PR DESCRIPTION
**Pull request by FpMarinov**
…ed period multiplied by 1.5, do a refactor on default retryer

### CODE REVIEW

**Review by velo — COMMENTED** [review]
Change makes sense, just needs to get build passing

**Comment by FpMarinov** [comment]
> Change makes sense, just needs to get build passing

This seems to be the cause of the problem, the weird thing is the test passes fine locally:

[ERROR] loadBalancingTargetPath(feign.ribbon.LoadBalancingTargetTest)  Time elapsed: 0.825 s  <<< ERROR!
feign.RetryableException: Connection refused executing GET http://localhost:53935/context-path/servers

at feign.ribbon.LoadBalancingTargetTest.loadBalancingTargetPath(LoadBalancingTargetTest.java:83)

Caused by: java.net.ConnectException: Connection refused


Do you have any ideas what could be going wrong?

**Comment by kdavisk6** [comment]
Per #1727

File: core/src/main/java/feign/Retryer.java [hunk 1]
```diff
@@ -47,35 +48,38 @@ public Default(long period, long maxPeriod, int maxAttempts) {
...
 
-    // visible for testing;
-    protected long currentTimeMillis() {
-      return System.currentTimeMillis();
-    }
-
+    @Override
     public void continueOrPropagate(RetryableException e) {
-      if (attempt++ >= maxAttempts) {
+      if (attempt >= maxAttempts) {
```
Comment (kdavisk6) [inline core/src/main/java/feign/Retryer.java:53]: This change is not backward compatible.  We count the attempt before sleeping intentionally.  This change will reduce the attempts by 1, please revert this change.

File: core/src/main/java/feign/Retryer.java [hunk 2]
```diff
@@ -47,35 +48,38 @@ public Default(long period, long maxPeriod, int maxAttempts) {
...
-        if (interval > maxPeriod) {
-          interval = maxPeriod;
-        }
+      if (e.retryAfter() == null) {
+        interval = nextMaxInterval();
+      } else {
+        interval = intervalFromRetryAfter(e.retryAfter());
+
         if (interval < 0) {
+          attempt++;
```
Comment (kdavisk6) [inline core/src/main/java/feign/Retryer.java:64]: See my previous comment about counting attempts.

File: core/src/main/java/feign/Retryer.java [hunk 3]
```diff
@@ -47,35 +48,38 @@ public Default(long period, long maxPeriod, int maxAttempts) {
...
+
       try {
         Thread.sleep(interval);
       } catch (InterruptedException ignored) {
         Thread.currentThread().interrupt();
         throw e;
       }
+
       sleptForMillis += interval;
+      attempt++;
```
Comment (kdavisk6) [inline core/src/main/java/feign/Retryer.java:77]: One more for the attempt count.

File: ribbon/src/test/java/feign/ribbon/RibbonClientTest.java [hunk 4]
```diff
@@ -339,7 +340,7 @@ public void testFeignOptionsNoFollowRedirect() {
     try {
       Response response = api.get();
       assertEquals(200, response.status());
-      assertEquals("Hello", response.body().toString());
+      assertEquals("Hello", new Scanner(response.body().asInputStream()).nextLine());
```
Comment (kdavisk6) [inline ribbon/src/test/java/feign/ribbon/RibbonClientTest.java:343]: There should be no need for a `Scanner` here.

Comment (kdavisk6) [reply ribbon/src/test/java/feign/ribbon/RibbonClientTest.java:343]: Also, this test is ignored.

File: core/src/test/java/feign/RetryerTest.java [hunk 5]
```diff
@@ -91,7 +91,7 @@ public void defaultRetryerFailsOnInterruptedException() {
       Assert.fail("Retryer continued despite interruption");
     } catch (RetryableException e) {
       Assert.assertTrue("Interrupted status not reset", Thread.interrupted());
-      Assert.assertEquals("Retry attempt not registered as expected", 2, retryer.attempt);
+      Assert.assertEquals("Retry attempt not registered as expected", 1, retryer.attempt);
```
Comment (kdavisk6) [inline core/src/test/java/feign/RetryerTest.java:94]: Last comment related to attempts.

**Review by kdavisk6 — CHANGES_REQUESTED** [review]
Changing how we count the number of attempts will break backward compatibility.  The original implementation is intentional.

**Pull request closed without merging**